third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -173,30 +173,31 @@ impl<'a> CyclicInstantiationChecker<'a> {
     /// Precondition: `callers_chain` is not empty
     fn report_error(
         &self,
-        _nid: NodeId,
+        nid: NodeId,
         callee: QualifiedInstId<FunId>,
         callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
     ) {
         let root = callers_chain[0].1.id;
         let mut labels = (0..callers_chain.len() - 1)
             .map(|i| {
-                let (caller_loc, caller) = &callers_chain[i];
+                let (_caller_loc, caller) = &callers_chain[i];
                 // callee of `caller`
-                let callee = &callers_chain[i + 1].1;
+                let (callee_loc, callee) = &callers_chain[i + 1];
                 format!(
                     "`{}` calls `{}` {}",
                     self.display_call(caller, root),
                     self.display_call(callee, root),
-                    caller_loc.display_line_only(self.mod_env.env)
+                    callee_loc.display_line_only(self.mod_env.env)
                 )
             })
             .collect_vec();
-        let (caller_loc, caller) = &callers_chain.last().expect("parent");
+        let (_caller_loc, caller) = &callers_chain.last().expect("parent");
+        let callee_loc = self.mod_env.env.get_node_loc(nid);
         labels.push(format!(
             "`{}` calls `{}` {}",
             self.display_call(caller, root),
             self.display_call(&callee, root),
-            caller_loc.display_line_only(self.mod_env.env)
+            callee_loc.display_line_only(self.mod_env.env)
```

Comment:
Could we use `error_with_labels` (or `diag_with_primary_notes_and_labels`) instead, where the location and associated code gets displayed much more nicely and provides more context?

Reply:
See reference safety processor for some examples.

Reply:
It is also an odd one out as far as error messages are concerned: `display_line_only` is not really used anywhere in the compiler core.

Reply:
I worry that this will cause the errors to increase in size too much.  This is comparable to V1's errors, see tests/move_check/translated_ir_tests/move/generics/instantiation_loops/mutually_recursive_three_args_type_con_shifting.exp

Reply:
We explained this in the original PR. The problem with `error_with_labels` is that we don't control the order of the error messages. They are shown in the order of source location, instead of the order in the call chains. The resulting error messages are disastrous. But once `error_with_labels` can preserve the order of error messages as we added them, we should switch to that to provide more context.